### ISSUE TITLE
Feature request: Auto JPA migration generation from "diff" 

### ISSUE THREAD

**Issue by mikehaas763** [flyway/flyway#648]
There is an awesome feature in Symfony and Doctrine/DoctrineMigrations, as well as with Rails migrations. I'll explain my work flow in past projects where I used doctrine to explain a feature I think would be absolutely awesome in flyway and would really give it a huge extra edge. It's great when working with entity/domain driven development. It wouldn't even have to be in core, just a well supported easily includable extension.

For those unfamiliar with these: 

Symfony = Spring/Rails/.NET MVC/etc
Doctrine = Hibernate (Very influenced)
Doctrine Migrations = Flyway/Liquibase

Doctrine Migrations from a dev user standpoint is pretty similar to flyway. Create a PHP class and add an up method (a down method also). However, they use a DateTime string on the file to indicate running order. It looks like Version20130417101156.php

Anyways, here is an example workflow:

In a Symfony app, I would create an entity, add some properties with the appropriate annotations such as for marking M:N relationships etc.

Then, I would use the Symfony "console" like so:

./app/console doctrine:migrations:diff

This would automatically create a migration file. It would do this by running a "diff" between the current DB and the mapping/annotation metadata on the entities. This would work perfectly 95% of the time. You are encouraged to modify the migration file if necessary.

I'm guessing it worked by just capturing the sql that doctrine would generate to create the schema.

I don't know if JPA specifies the api for accessing the sql that would be created from an implementation. If so, that would be great something could be written to cover them all. If not, sucks but gotta start somewhere (I would start with hibernate's api).

Is anyone else interested in something like this? It would be nice to get a group of community members working on something like this. Not everyone would like this type of workflow but for those that do and would, it is a huge time saver. Add properties to classes and avoid creating any files/adding any boilerplate/and writing any sql.

**Comment by velikanov** [comment]
@axelfontaine are you serious? Why?
the automatic diff is the greatest time-saver (in Symfony for sure) feature

**Comment by mikehaas763** [comment]
@velikanov I don't think that he's saying no one would be interested in having the feature, just that no one had been interested in implementing it. I'm sure it's a decently large undertaking.

**Comment by velikanov** [comment]
@mikehaas763, I got it, anyway I'm looking at liquibase now

**Comment by axelfontaine** [comment]
In 12 months nobody +1 the issue. In the end it is all about priorities, based on the demands of the community. Feel free to submit a PR if you want to do something about this.

**Comment by velikanov** [comment]
:+1: :)
I'm hardly new to Java, so I just can not do this in sane time :(

**Comment by roblight** [comment]
+1

Would love to see JHipster support Flyway (for reference: https://github.com/jhipster/generator-jhipster/issues/2019).

**Comment by rcomblen** [comment]
+1

Having experience with SQLAlchemy and Alembic in the Python world, this is the first feature I want from a tool like Flyway: autogenerate the migration script based on the ORM information.

**Comment by chence** [comment]
+1

**Comment by carrchang** [comment]
+1

**Comment by kdazzle** [comment]
+1

Something like Django's migrations would be amazing

**Comment by apxboxfuse** [comment]
+1 I too would love to see JHipster support flyway

**Comment by frankenpfalzer** [comment]
+1
I'm new to Flyway but that Feature would be a real time saver.

**Comment by mustache1up** [comment]
+1

**Comment by carrchang** [comment]
I am using JOOQ and Spring JdbcTemplate to implement Flyway Java migration, add a gist [here](https://gist.github.com/carrchang/8b32df7a76f7e9738bf228601e6819e8), may be a little help.

**Comment by nicwillemse** [comment]
perhaps this should be reopened ?

**Comment by YouYue123** [comment]
+1

**Comment by idchlife** [comment]
I'm sorry, but there is no auto-generate migration with diffing existing schema with entities in flyway?

**Comment by jkubrynski** [comment]
Hi,

we've implemented a plugin which generates the Flyway schema migrations based on JPA model: https://github.com/Devskiller/jpa2ddl

**Comment by lewandom** [comment]
+1

**Comment by jkubrynski** [comment]
I've created a sample project with generating Flyway migrations based on the JPA model:
https://github.com/Devskiller/jpa2ddl/tree/master/jpa2ddl-samples/jpa2ddl-flyway-maven-sample

**Comment by marcin-chwedczuk** [comment]
+1

**Comment by TorstenCScholz** [comment]
+1

**Comment by jcharlytown** [comment]
+1

**Comment by riaprocoumai** [comment]
+1

**Comment by Vynny** [comment]
+1

**Comment by cbou** [comment]
All feature with generation e.g. like #400 help new user to jump in the project and get more confident with it.

Still I'm not sure it has to be part of Flyway actually. A diff tools would need to understand SQL, JPA or a schema (like Doctrine) which is a lot more than what Flyway currently knows.

To me such a feature would better fit in Hibernate.

**Comment by unknowedNp** [comment]
+1

**Comment by mohsenr7596** [comment]
+1

**Comment by gediminas-metasite** [comment]
+1

**Comment by mohoma** [comment]
+1

**Comment by niekvangogh** [comment]
+1

**Comment by arianofaoms** [comment]
+1

**Comment by doublemcz** [comment]
+1

**Comment by pflorek** [comment]
If we are using hibernate, this feature could be a duplicate of hibernate's possibilities. Since hibernate comes with it's own SchemaManagementTool, it migrates the db or writes the migration into a file: [automatic schema generation](https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#configurations-hbmddl)

If spring is used, properties needs to be prefixed with **spring.jpa.properties**:

```properties
spring.jpa.properties.hibernate.format_sql=false
spring.jpa.properties.hibernate.hbm2ddl.delimiter=;
spring.jpa.properties.javax.persistence.schema-generation.create-source=metadata
spring.jpa.properties.javax.persistence.schema-generation.database.action=validate
spring.jpa.properties.javax.persistence.schema-generation.scripts.action=update
spring.jpa.properties.javax.persistence.schema-generation.scripts.create-target=src/main/resources/db/migration/Vx__New_Migration.sql
spring.jpa.properties.javax.persistence.schema-generation.scripts.drop-target=src/main/resources/db/migration/Vx__New_Migration.sql
```

Only drawback is, you need to bump the version yourself 😉 

I also tried @jkubrynski 's great [jpa2ddl](https://github.com/Devskiller/jpa2ddl). But didn't managed to load the current migrations into the memory database, resulting in a full schema drop. Which is nearly same to `javax.persistence.schema-generation.scripts.action=create`. Drawback was using a in mem db.

So i played a lil bit with jpa2ddl and [hibernate integrator](https://docs.jboss.org/hibernate/orm/5.4/integrationguide/html_single/Hibernate_Integration_Guide.html). Just extended [org.hibernate.integrator.spi.Integrator](https://docs.jboss.org/hibernate/orm/5.4/javadocs/org/hibernate/integrator/spi/Integrator.html) and added that class reference to `/META-INF/services/org.hibernate.integrator.spi.Integrator` to let hibernate load that class on bootstrap. There I got access to `org.hibernate.boot.Metadata`, while hibernate was setting up `org.hibernate.internal.SessionFactoryImpl` before schema tool was starting it's work.

My conclusion, hibernates schema tool is almost sufficient. We just should a add a short instruction for hibernate users to flyway docs, how to create an update migration ;) Would have saved me a lot of time deep diving into hibernate, since their docs aren't that precise:

> javax.persistence.schema-generation.scripts.action (e.g. none (default value), create-only, drop, create, create-drop, validate, and update)
Setting to perform SchemaManagementTool actions writing the commands into a DDL script file. Valid options are defined by the externalJpaName value of the Action enum:
none
No action will be performed.
create
Database creation will be generated.
drop
Database dropping will be generated.
drop-and-create
Database dropping will be generated followed by database creation.

**Comment by andreyoganesyan** [comment]
This can be achieved with the [JPA Buddy](https://plugins.jetbrains.com/plugin/15075-jpa-buddy) plugin for IntelliJ IDEA. It has a feature that generates Flyway migration scripts by:
* Finding the difference between JPA entities and the database
* Comparing two databases representing current and legacy states

This is what it looks like: https://www.youtube.com/watch?v=O16D86Zvh-o

**Comment by DoodleBobBuffPants** [comment]
Closing as it looks like this can be achieved with JPA Buddy. Can reopen if there's a need for it as a Flyway capability

**Comment by daniel-frak** [comment]
I don't think that a plugin for an IDE can be considered a valid solution. One should never depend on using a specific IDE (let alone a third party plugin for it) to achieve things like this.

IMO the issue should be reopened.
It would be great if Flyway had this capability natively.

**Comment by WalterMourao** [comment]
Hi @pflorek thank you the clever hibernate configuration. The drawback in using the hibernate schema generation is the lack of element removal. Hibernate checks if the JPA entities fit the schema, but it does not check if some element has been removed/changed.

**Comment by pflorek** [comment]
> Hi @pflorek thank you the clever hibernate configuration. The drawback in using the hibernate schema generation is the lack of element removal. Hibernate checks if the JPA entities fit the schema, but it does not check if some element has been removed/changed.

@WalterMourao 

As far as I remember I tested the `create-target` and it was fine for my use case. What exactly is missing in the `drop-target`

**Comment by Dzeri96** [comment]
> I don't think that a plugin for an IDE can be considered a valid solution. One should never depend on using a specific IDE (let alone a third party plugin for it) to achieve things like this.
> 
> IMO the issue should be reopened. It would be great if Flyway had this capability natively.

Not to mention that this is now a premium function in JPA Buddy, meaning you have to pay for it

**Comment by danielsawan** [comment]
10 years that peoples are waiting for this feature to be implemented. what a joke.
here is an alternative : https://contribute.liquibase.com/extensions-integrations/directory/database-tutorials/hibernate/
